### PR TITLE
add ignore_class param to sparse CE loss

### PR DESCRIPTION
**Pull request by AakashKumarNain**
Add `ignore_class` param for sparse categorical cross entropy loss

### CODE REVIEW

**Comment by AakashKumarNain** [comment]
@fchollet `ops.where` is pretty inconsistent across the backends, and within a single backend itself. The bigger problem though is that it accepts two positional arguments which aren't **optional**. Ideally it should work just with the condition. This is the reason why tests for `torch` is failing

**Comment by fchollet** [comment]
> Do we make it consistent across the backends or should we leave it to the behavior of the backend?

We should make the behavior consistent across all backends -- that's important. And when applicable the behavior should also be standardized on the numpy behavior.

**Comment by AakashKumarNain** [comment]
@fchollet I have skipped `torch` backend for now. `ops.scatter(...)`  in torch backend seems broken, and I have no idea how to fix it as of now.

**Comment by AakashKumarNain** [comment]
I just realized that even though the tests for jax are passing, I am skeptical about it working when the loss function is complied. Why? Because the masking is dynamic in this case

File: keras_core/losses/losses.py [hunk 1]
```diff
@@ -1682,13 +1693,45 @@ def sparse_categorical_crossentropy(y_true, y_pred, from_logits=False, axis=-1):
...
+    res = ops.sparse_categorical_crossentropy(
         y_true,
         y_pred,
         from_logits=from_logits,
         axis=axis,
     )
 
+    if ignore_class is not None:
+        if backend.backend() == "jax":
+            res_shape = tuple(ops.cast(y_pred_shape[:-1], "int32"))
```
Comment (fchollet) [inline keras_core/losses/losses.py:1719]: You can use `int` to specify a default backend int type

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. What's the error in torch? And for JAX, surely correctness is being tested so the test passing should be proof that it works?

**Comment by AakashKumarNain** [comment]
Thanks for review it. Here is the traceback for torch:

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)


File ~/keras-core/keras_core/losses/loss.py:44, in Loss.__call__(self, y_true, y_pred, sample_weight)
     37 y_pred = tree.map_structure(
     38     lambda x: ops.convert_to_tensor(x, dtype=dtype), y_pred
     39 )
     40 y_true = tree.map_structure(
     41     lambda x: ops.convert_to_tensor(x, dtype=dtype), y_true
     42 )
---> 44 losses = self.call(y_true, y_pred)
     45 out_mask = getattr(losses, "_keras_mask", None)
     47 if in_mask is not None and out_mask is not None:

Cell In[5], line 20, in LossFunctionWrapper.call(self, y_true, y_pred)
     18 def call(self, y_true, y_pred):
     19     y_true, y_pred = squeeze_to_same_rank(y_true, y_pred)
---> 20     return self.fn(y_true, y_pred, **self._fn_kwargs)

Cell In[13], line 57, in sparse_categorical_crossentropy(y_true, y_pred, from_logits, ignore_class, axis)
     54 if isinstance(valid_idx, (tuple, list)):
     55     valid_idx = ops.stack(valid_idx, -1)
---> 57 res = ops.scatter(indices=valid_idx, values=res, shape=res_shape)
     59 if backend.backend() != "numpy":
     60     res._keras_mask = valid_mask

File ~/keras-core/keras_core/ops/core.py:61, in scatter(indices, values, shape)
     59 if any_symbolic_tensors((indices, values, shape)):
     60     return Scatter().symbolic_call(indices, values, shape)
---> 61 return backend.core.scatter(indices, values, shape)

File ~/keras-core/keras_core/backend/torch/core.py:277, in scatter(indices, values, shape)
    275 value_shape = shape[index_length:]
    276 indices = torch.reshape(indices, [-1, index_length])
--> 277 values = torch.reshape(values, [-1] + list(value_shape))
    279 for i in range(indices.shape[0]):
    280     index = indices[i]

RuntimeError: shape '[-1, 2]' is invalid for input of size 1
```

> And for JAX, surely correctness is being tested so the test passing should be proof that it works?

I don't know how the test is passing because I definitely get error when I try to `jit` it separately. And the error makes sense as well. This is the version that is jittable in JAX and works as expected

```python
def loss_fn(y_true, y_pred, ignore_class=-1):
    mask = ops.equal(y_true, ignore_class)
    mask_idx = ops.stack(jnp.nonzero(mask, size=len(mask)), -1)
    
    y_true_masked = ops.where(mask, 0.0, y_true)
    y_pred_masked = ops.where(mask, 0.0, y_pred)
    
    loss = ops.sparse_categorical_crossentropy(
        target=y_true_masked,
        output=y_pred_masked,
        from_logits=True,
        axis=-1
    )
    
    mask = ops.reshape(mask, y_pred.shape[:-1])
    loss = ops.where(mask, 0.0, loss)
    return loss
```

The problem with this code though that it implements the lax compatible `nonzero(...)` method that takes the `size` argument. This argument is missing from other backends. One way to tackle this problem is to handle it in a `cond` depending on the backend, and implement two version of it, one for JAX, and one for others

**Comment by kerighan** [comment]
I don't know if it's the place to add a comment or try to help and I may be mistaken, but I believe the behaviour of ignore_class could be expected without using ops.where, for example using the following implementation:

`
def sparse_categorical_crossentropy_loss(ignore_class=None):
    def wrapper(y_true, y_pred_logits):
        # Transforming logits to softmax for probability distribution
        y_pred = ops.nn.softmax(y_pred_logits, axis=-1)

        # Getting the number of classes from predictions
        num_classes = ops.shape(y_pred)[-1]

        # One-hot encoding the true labels
        y_true_one_hot = ops.one_hot(
            ops.cast(y_true, "int32"), num_classes)

        # Mask for valid positions (not ignore_class), if specified
        if ignore_class is not None:
            mask = ops.not_equal(y_true, ignore_class)
            mask = ops.cast(mask, "float32")
            y_true_one_hot *= ops.expand_dims(mask, axis=-1)
            # Masking the one-hot labels

        # Adding a small value to avoid computing log(0) which gives NaN
        y_pred = ops.clip(y_pred, 1e-7, 1.)

        # Computing cross-entropy
        loss = -ops.sum(y_true_one_hot * ops.log(y_pred), axis=-1)

        return ops.mean(loss)
    return wrapper
`
It does not use ops.sparse_categorical_crossentropy though, but the masking bit seems to work accross all backends.

**Comment by AakashKumarNain** [comment]
Thanks for the suggestion. If you look at the hanged files, this is exactly what it does except for using `ops.sparse_categoical_crossentropy(...)`, and the reason for that is we avoid redundancy as much as possible.

**Comment by AakashKumarNain** [comment]
Update: Now that I have fixed the `where` op, I have a version that works with all the backends. The only catch is with JAX, as the version won't be jittable. Figuring out a way to do that as well

**Comment by AakashKumarNain** [comment]
@fchollet good news! The current version is now fully jit compatible.

File: keras_core/losses/losses.py [hunk 2]
```diff
@@ -1659,14 +1663,21 @@ def categorical_focal_crossentropy(
...
 
     Args:
         y_true: Ground truth values.
         y_pred: The predicted values.
         from_logits: Whether `y_pred` is expected to be a logits tensor. By
             default, we assume that `y_pred` encodes a probability distribution.
+        ignore_class: Optional integer. The ID of a class to be ignored during
+            loss computation. This is useful, for example, in segmentation
+            problems featuring a "void" class (commonly -1 or 255) in
+            segmentation maps. By default (ignore_class=None), all classes are
```
Comment (fchollet) [inline keras_core/losses/losses.py:1679]: Use backticks around code keywords

File: keras_core/losses/losses.py [hunk 3]
```diff
@@ -1682,13 +1693,31 @@ def sparse_categorical_crossentropy(y_true, y_pred, from_logits=False, axis=-1):
...
         y_pred,
         from_logits=from_logits,
         axis=axis,
     )
 
+    if ignore_class is not None:
+        valid_mask = ops.reshape(valid_mask, res_shape)
+        res = ops.where(valid_mask, res, 0.0)
+
+        if backend.backend() != "numpy":
```
Comment (fchollet) [inline keras_core/losses/losses.py:1716]: Rather than using this conditional, do a try/except AttributeError.

**Review by fchollet — COMMENTED** [review]
Thanks for the update! Looking good

File: keras_core/losses/losses.py (hunk 2 above)
Comment (AakashKumarNain) [reply keras_core/losses/losses.py:1679]: Sure.

File: keras_core/losses/losses.py (hunk 3 above)
Comment (AakashKumarNain) [reply keras_core/losses/losses.py:1716]: If we use `try/except`, and use a print statement inside it, the function will have a side-effect when jit compiled. Is there any specific reason for not using `if` here?

Comment (fchollet) [reply keras_core/losses/losses.py:1716]: Try/except is not a side effect and there is no print statement?

Comment (AakashKumarNain) [reply keras_core/losses/losses.py:1716]: Okay, so I misunderstood. I thought we need to add a message as well. Thanks for the clarification. I will make the change

**Comment by AakashKumarNain** [comment]
@fchollet thank you for the review and the valuable suggestions. I have made the necessary changes.

**Review by fchollet — APPROVED** [review]
LGTM, thanks!

**Pull request merged**